test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/ModelReferenceTests/ModelReferenceClientTests.cs

```diff
@@ -0,0 +1,426 @@
+        newCreated.VehicleSpeed = 200.1;
+        context.UpdateObject(newCreated);
+        await context.SaveChangesAsync();
+
+        // Query and Delete entity
+        VehicleGPSType updated = queryable.Single();
+        Assert.Equal(200.1, newCreated.VehicleSpeed);
+
+        context.DeleteObject(updated);
+        await context.SaveChangesAsync();
+        Assert.Equal(3, vehicleGPSSet.Count());
+    }
+
+    // Query set - Create entity - Get created entity - Update entity - Delete entity of set in main entity container
+    // Type defined in main container, but derived from type referenced model
+    [Fact]
+    public void QueryCreateUpdateDelete_EntitySetDerivedFromReferencedModelType()
+    {
+        // Arrange
+        var context = this.ContextWrapper();
+        context.MergeOption = MergeOption.OverwriteChanges;
+
+        // Query VehicleGPSSet
+        var derivedVehicleGPSSet = context.DerivedVehicleGPSSet;
+        Assert.Equal(2, derivedVehicleGPSSet.Count());
+
+        // Create an entity in VehicleGPSSet
+        var newVehicleGPS = new DerivedVehicleGPSType()
+        {
+            Key = "101",
+            VehicleSpeed = 100.1,
+            StartLocation = new GeoLocation()
+            {
+                Lat = 1,
+                Long = 2,
+            },
+            EndLocation = new GeoLocation()
+            {
+                Lat = 3,
+                Long = 4,
+            },
+            CurrentLocation = new GeoLocation()
+            {
+                Lat = 1.2,
+                Long = 2.4,
+            },
+            LostSignalAlarm = new GPSLostSignalAlarmType()
+            {
+                Severity = 1,
+                LastKnownLocation = new GeoLocation()
+                {
+                    Lat = 2.1,
+                    Long = 1.2,
+                }
+            },
+            Map = new MapType()
+            {
+                MBytesDownloaded = 1.2,
+                ProviderName = "TESTNEW",
+                Uri = "TESTNEW.TEST",
+            },
+            DisplayName = "TESTTEST"
+        };
+
+        context.AddToDerivedVehicleGPSSet(newVehicleGPS);
+        context.SaveChanges();
+
+        // Get the created entity
+        var queryable = context.DerivedVehicleGPSSet.Where(vehicleGPS => vehicleGPS.Key == "101");
+        VehicleGPSType newCreated = queryable.Single();
+        Assert.Equal(100.1, newCreated.VehicleSpeed);
+
+        // Update the created entity 
+        newCreated.VehicleSpeed = 200.1;
+        context.UpdateObject(newCreated);
+        context.SaveChanges();
+
+        // Query and Delete entity
+        VehicleGPSType updated = queryable.Single();
+        Assert.Equal(200.1, newCreated.VehicleSpeed);
+
+        context.DeleteObject(updated);
+        context.SaveChanges();
+        Assert.Equal(2, derivedVehicleGPSSet.Count());
+    }
+
+    [Fact]
+    public void QueryPropertiesAndNavigationProperties_FromMainAndReferencedModels()

```

Comment:
Can these tests be broken into different tests. Have tests for loading properties from main models. And have separate tests for loading properties from referenced models.